runtime/registry.go

```diff
@@ -79,11 +72,148 @@ func (r *Runtime) CreateInstance(ctx context.Context, inst *drivers.Instance) er
+	olapChanged := olderInstance.OLAPDriver != inst.OLAPDriver || olderInstance.OLAPDSN != inst.OLAPDSN
+	if olapChanged {
+		// Check OLAP connection
+		olap, _, err := r.checkOlapConnection(inst)
+		if err != nil {
+			return err
+		}
+
+		// Prepare connections for use
+		err = olap.Migrate(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to prepare instance: %w", err)
+		}
+		olapConn = olap
+	}
+
+	// 2. embedCatalog disabled previously but enabled now
+	if inst.EmbedCatalog {
+		if olapConn == nil {
+			// getting exisiting connection
+			olapConn, err = r.connCache.get(ctx, inst.ID, inst.OLAPDriver, inst.OLAPDSN)
+			if err != nil {
+				return err
+			}
+		}
+		_, ok := olapConn.CatalogStore()
+		if !ok {
+			return errors.New("driver does not support embedded catalogs")
+		}
+	}
+
+	// 3. changes in repo driver or repo dsn
+	repoChanged := inst.RepoDriver != olderInstance.RepoDriver || inst.RepoDSN != olderInstance.RepoDSN
+	if repoChanged {
+		// Check repo connection
+		repo, _, err := r.checkRepoConnection(inst)
+		if err != nil {
+			return err
+		}
+
+		// Prepare connections for use
+		err = repo.Migrate(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to prepare instance: %w", err)
+		}
+	}
+
+	// evict caches if connections need to be updated
+	if olapChanged || repoChanged {
+		r.evictCaches(ctx, olderInstance)
+	}
+
+	// update env variables
+	if inst.Env == nil {
+		inst.Env = make(map[string]string)
+	}
+	inst.Env["allow_host_credentials"] = strconv.FormatBool(r.opts.AllowHostCredentials)
+	// update the entire instance for now to avoid building queries in some complicated way
+	return r.Registry().EditInstance(ctx, inst)
+}
+
+// TODO :: this is a rudimentary solution and ideally should be done in some producer/consumer pattern
+func (r *Runtime) evictCaches(ctx context.Context, inst *drivers.Instance) {
+	// evict and close exisiting connection
+	r.connCache.evict(ctx, inst.ID, inst.OLAPDriver, inst.OLAPDSN)
+	r.connCache.evict(ctx, inst.ID, inst.RepoDriver, inst.RepoDSN)
+
+	// evict catalog cache
+	r.catalogCache.evict(ctx, inst.ID)
+	// query cache can't be evicted since key is a combination of instance ID and other parameters
+}
+
+func (r *Runtime) checkRepoConnection(inst *drivers.Instance) (drivers.Connection, drivers.RepoStore, error) {
+	repo, err := drivers.Open(inst.RepoDriver, inst.RepoDSN, r.logger)
+	if err != nil {
+		return nil, nil, err
+	}
+	repoStore, ok := repo.RepoStore()
+	if !ok {
+		return nil, nil, fmt.Errorf("not a valid repo driver: '%s'", inst.RepoDriver)
+	}
+
+	return repo, repoStore, nil
+}
+
+func (r *Runtime) checkOlapConnection(inst *drivers.Instance) (drivers.Connection, drivers.OLAPStore, error) {
+	olap, err := drivers.Open(inst.OLAPDriver, inst.OLAPDSN, r.logger)
+	if err != nil {
+		return nil, nil, err
+	}
+	olapStore, ok := olap.OLAPStore()
+	if !ok {
+		return nil, nil, fmt.Errorf("not a valid OLAP driver: '%s'", inst.OLAPDriver)
+	}
+	return olap, olapStore, nil
 }
```

Comment:
These should a) run migrate, b) close the connection again (since not caching here).

But a better design might be to run the check through the cache, to prevent opening/closing unnecessarily. Don't have to do that in this PR though, given other plans

Reply:
I wasn't sure if I should combine the open connection with migrate. I didn't close the connection because of the in-memory duck db constraint and test case fixes. 
But agreed all of this will be better handled through the cache in other refactor PR.